source/adapters/opencl/context.hpp

```diff
@@ -10,9 +10,74 @@
 #pragma once
 
 #include "common.hpp"
+#include "device.hpp"
 
-namespace cl_adapter {
-ur_result_t
-getDevicesFromContext(ur_context_handle_t hContext,
-                      std::unique_ptr<std::vector<cl_device_id>> &DevicesInCtx);
-}
+#include <vector>
+
+struct ur_context_handle_t_ {
+  using native_type = cl_context;
+  native_type Context;
+  std::vector<ur_device_handle_t> Devices;
+  uint32_t DeviceCount;
+  std::atomic<uint32_t> RefCount = 0;
+
+  ur_context_handle_t_(native_type Ctx, uint32_t DevCount,
+                       const ur_device_handle_t *phDevices)
+      : Context(Ctx), DeviceCount(DevCount) {
+    for (uint32_t i = 0; i < DeviceCount; i++) {
+      Devices.emplace_back(phDevices[i]);
+      urDeviceRetain(phDevices[i]);
+    }
+    urContextRetain(this);
+  }
+
+  uint32_t incrementReferenceCount() noexcept { return ++RefCount; }
+
+  uint32_t decrementReferenceCount() noexcept { return --RefCount; }
+
+  uint32_t getReferenceCount() const noexcept { return RefCount; }
+
+  static ur_result_t makeWithNative(native_type Ctx, uint32_t DevCount,
+                                    const ur_device_handle_t *phDevices,
+                                    ur_context_handle_t &Context) {
+    try {
+      auto URContext =
+          std::make_unique<ur_context_handle_t_>(Ctx, DevCount, phDevices);
+      native_type &NativeContext = URContext->Context;
+      uint32_t &DeviceCount = URContext->DeviceCount;
+      if (!DeviceCount) {
+        CL_RETURN_ON_FAILURE(
+            clGetContextInfo(NativeContext, CL_CONTEXT_NUM_DEVICES,
+                             sizeof(DeviceCount), &DeviceCount, nullptr));
+        std::vector<cl_device_id> CLDevices(DeviceCount);
+        CL_RETURN_ON_FAILURE(clGetContextInfo(NativeContext, CL_CONTEXT_DEVICES,
+                                              sizeof(CLDevices),
+                                              CLDevices.data(), nullptr));
+        URContext->Devices.resize(DeviceCount);
+        for (uint32_t i = 0; i < DeviceCount; i++) {
+          ur_native_handle_t NativeDevice =
+              reinterpret_cast<ur_native_handle_t>(CLDevices[i]);
+          UR_RETURN_ON_FAILURE(urDeviceCreateWithNativeHandle(
```

Comment:
do we need to do this here? I thought the application would already have had to do something similar to get the devices for `phDevices`

Reply:
Some e2e tests (especially opencl interop ones) where creating ctx from native context but passing devices as `nullptr`, then when it queries the context handle for the devices, it will create a problem as it will expect UR devices handles related to the context, that will also have the native handle for devices and it didn't pass them in first place, so I need to create ones for it. (In OpenCL Interop functionality, it could create any handle at any point without having to pass the dependant handles so thats why I need to create ones, and that is mostly as also SYCl runtime was built to cast the OPENCL handles directly, so i also had to make some changes there to align with that, the changes are in the testing PR for this PR).

Reply:
passing nullptr for `phDevices` isn't allowed by the spec, so this needs a spec fix or a sycl fix

Reply:
I think it needs a spec fix as I don't find a reference for it in [here](https://oneapi-src.github.io/unified-runtime/core/api.html#urcontextcreatewithnativehandle)

Reply:
I also think, it might need changes in the e2e tests/sycl-rt

Reply:
Ah okay, I didn't notice it in the spec. I think sycl-rt maybe has something to change for that, will try to retest without this and collect more info on this part.

Reply:
I can't remember why we added the device list as a non-optional parameter but I do remember that it happened relatively recently, maybe we should re-assess that if it doesn't align with the sycl spec. @kbenzie can you remember what the deal with this was?

Reply:
They were added in #433 to line up with the PI interface which was added in https://github.com/intel/llvm/pull/2772. Both motivated by Level Zero interop.

Reply:
I moved that logic to intel/llvm side so that we don't need to change any spec.